src/redprl/syntax.sml

```diff
@@ -67,17 +82,45 @@ struct
        | PATH_ABS (u, m) => O.MONO O.PATH_ABS $$ [([u],[]) \ m]
        | PATH_AP (m, r) => O.POLY (O.PATH_AP r) $$ [([],[]) \ m]
 
-       | HCOM {dir, class=ty, cap, tubes} =>
+       | FHCOM {dir, cap, tubes} =>
            let
-             val (eqs, tubes) = ListPair.unzip tubes
-             val tubes' = List.map (fn (d, t) => ([d], []) \ t) tubes
+             val (eqs, tubes) = tubesInto tubes
            in
-             O.POLY (O.HCOM (O.TAG_NONE, eqs, dir)) $$
-               (([],[]) \ ty) :: (([],[]) \ cap) :: tubes'
+             O.POLY (O.FHCOM (dir, eqs)) $$ (([],[]) \ cap) :: tubes
            end
        | CUST => raise Fail "CUST"
        | META => raise Fail "META"
 
+    val intoAp = into o AP
+    val intoLam = into o LAM
+
+    val intoFst = into o FST
+    val intoSnd = into o SND
+    val intoPair = into o PAIR
+
+    fun intoHcom' (dir, eqs) (ty, args) =
+      O.POLY (O.HCOM (dir, eqs)) $$ (([],[]) \ ty) :: args
+
+    fun intoHcom (dir, eqs) (ty, cap, tubes) =
+      intoHcom' (dir, eqs) (ty, (([],[]) \ cap) :: tubes)
+
+    fun intoFHcom' (dir, eqs) args = O.POLY (O.FHCOM (dir, eqs)) $$ args
+
+    fun intoFHcom (dir, eqs) (cap, tubes) =
+      intoFHcom' (dir, eqs) ((([],[]) \ cap) :: tubes)
+  
+    fun intoCoe dir ((u, a), m) =
+      O.POLY (O.COE dir) $$ [([u],[]) \ a, ([],[]) \ m]
+
+    fun intoCom (dir as (r, r'), eqs) ((u, a), cap, tubes) =
+      let
+        fun coe v m = intoCoe (v, r') ((u, a), m)
+        fun goTube (([v],_) \ n) = ([v],[]) \ coe (P.ret v) n
+          | goTube _ = raise Fail "malformed tube"
+      in
+        intoHcom (dir, eqs) (substSymbol (r', u) a, coe r cap, List.map goTube tubes)
+      end
```

Comment:
@favonia Yeah... I guess the `Tac` things are in `Machine` because they aren't used anywhere else, and there's no reason to expect that they'd be used anywhere else. But if you prefer / find it more systematic, you can feel free to move any helpers that you want into `Syntax`.

I think that there's not yet much rhyme nor reason to how things are arranged, so if you find something that you wish were elsewhere, you can feel free to move it.